crates/wasmtime/src/runtime/vm/component.rs

```diff
@@ -745,6 +745,75 @@ impl ComponentInstance {
         _ = (src_idx, src, dst);
         todo!()
     }
+
+    #[cfg(feature = "threads")]
+    pub(crate) fn thread_spawn_indirect(
+        &mut self,
+        table: u32,
+        element: u32,
+        _context: u32,
+    ) -> Result<u32> {
+        use crate::vm::{Instance, TableElement};
+        use crate::{Func, ValType};
+        use core::iter;
+
+        // Retrieve the table referenced by the canonical builtin (i.e.,
+        // `table`). By validation this is guaranteed to be a `shared funcref`
+        // table.
+        let VMTable { vmctx, from } = self.runtime_table(RuntimeTableIndex::from_u32(table));
+        let element = u64::from(element);
+        let table = unsafe {
+            Instance::from_vmctx(vmctx.as_non_null(), |handle| {
+                let idx = handle.table_index(from.as_non_null().as_ref());
+                handle
+                    .get_defined_table_with_lazy_init(idx, iter::once(element))
+                    .as_ref()
+            })
+        };
```

Comment:
FWIW technically some and/or most of this is going to go away. Lazy initialization is probably not going to be possible with shared tables.

Reply:
Yup. What I'm trying to do is plumb through enough to (a) show how this is currently possible using unshared tables (while still immediately failing) and (b) create a use site here so that future refactoring considers this usage. I was thinking that tests could even check for the initial behavior that _is_ functional today: i.e., we can't run a shared function, but at least we can check that using the wrong index results in a failure.

Reply:
I don't think it's possible to hit this at runtime though? All components would fail validation unless they use shared tables, and shared talbes aren't implemented in Wasmtime, so it's not possible to instantiate any component using this intrinsic?